app/models/assignment.rb

```diff
@@ -851,7 +853,72 @@ def get_num_valid
              .count(&:is_valid?)
   end
 
-  def get_num_marked(ta_id = nil)
+  def marked_result_ids_for(ta_id)
+    cache_ta_results.dig(ta_id, :marked_result_ids) || []
+  end
+
+  def cache_ta_results
+    return @cache_ta_results if defined? @cache_ta_results
+
+    data = current_results.joins(grouping: :tas).pluck('tas.id', 'results.id', 'results.marking_state')
+    # Group results by TA ID
+    grouped_data = data.group_by { |ta_id, _result_id, _marking_state| ta_id }
+
+    results_by_id = current_results.includes(:marks).index_by(&:id)
+
+    ta_to_criteria = self.criterion_ta_associations.group_by(&:ta_id).transform_values do |assocs|
```

Comment:
It's a bit cleaner to do a `.pluck([:ta_id, :criterion_id])` and then use a `group_by`, like you're doing for `grouped_data` above. This prevents a bunch of `CriterionTaAssociation` model objects from being created.